### PR TITLE
Do cover in another way

### PR DESCRIPTION
**Pull request by lyyyuna**

### CODE REVIEW

File: cmd/build_test.go [hunk 1]
```diff
@@ -108,12 +108,12 @@ func TestBuildBinaryForInternalPackage(t *testing.T) {
...
 
-	cmd := exec.Command("go", "tool", "objdump", "simple-project")
-	cmd.Dir = workingDir
-	out, _ := cmd.CombinedOutput()
-	cnt := strings.Count(string(out), "GoCacheCover")
-	assert.Equal(t, cnt > 0, true, "GoCacheCover variable for internal package should be in the binary")
-
-	cnt = strings.Count(string(out), "internal.GoCover")
-	assert.Equal(t, cnt > 0, true, "internal.GoCover variable should be in the binary")
+	// cmd := exec.Command("go", "tool", "objdump", "simple-project")
```
Comment (CarlJi) [inline cmd/build_test.go:111]: no need to keep the useless code

File: pkg/cover/cover.go [hunk 2]
```diff
@@ -142,27 +174,30 @@ func Execute(args, newGopath, target, mode, agentPort, center string) error {
...
 			log.Printf("handle package: %v", pkg.ImportPath)
 			// inject the main package
-			mainCover, err := AddCounters(pkg, mode, newGopath)
-			if err != nil {
-				log.Errorf("failed to add counters for pkg %s, err: %v", pkg.ImportPath, err)
-				return ErrCoverPkgFailed
-			}
+			mainCover, mainDecl, _ := addCounters2(pkg, mode, globalCoverVarImportPath)
+			allDecl += mainDecl
+			// if err != nil {
```
Comment (CarlJi) [inline pkg/cover/cover.go:185]: no need to keep this.

File: pkg/cover/cover.go [hunk 3]
```diff
@@ -315,6 +286,24 @@ func AddCounters(pkg *Package, mode, newgopath string) (*PackageCover, error) {
 	}, nil
 }
 
+// addCounters2 is different from official go tool cover
+// 1. only inject covervar++ into source file
+// 2. no declarartions for these covervars
+// 3. return the declarations as string
+func addCounters2(pkg *Package, mode string, globalCoverVarImportPath string) (*PackageCover, string, error) {
```
Comment (CarlJi) [inline pkg/cover/cover.go:293]: if `AddCounters` will not be used again, we can delete this func.

File: pkg/cover/cover2tool/cover.go [hunk 4]
```diff
@@ -0,0 +1,775 @@
...
+// var profile string // The profile to read; the value of -html or -func
+
+var counterStmt func(*File, string) string
+
+const (
+	atomicPackagePath = "sync/atomic"
+	atomicPackageName = "_cover_atomic_"
+)
+
+// func main() {
```
Comment (CarlJi) [inline pkg/cover/cover2tool/cover.go:68]: no need to keep these

File: pkg/cover/cover_test.go [hunk 5]
```diff
@@ -362,20 +370,29 @@ func TestCoverResultForInternalPackage(t *testing.T) {
...
 		assert.FailNow(t, "should generate http_cover_apis_auto_generated.go")
 	}
 
-	out, err := ioutil.ReadFile(filepath.Join(testDir, "http_cover_apis_auto_generated.go"))
-	if err != nil {
-		assert.FailNow(t, "failed to read http_cover_apis_auto_generated.go file")
-	}
-	cnt := strings.Count(string(out), "GoCacheCover")
-	assert.Equal(t, cnt > 0, true, "GoCacheCover variable should be in http_cover_apis_auto_generated.go")
+	// out, err := ioutil.ReadFile(filepath.Join(testDir, "http_cover_apis_auto_generated.go"))
```
Comment (CarlJi) [inline pkg/cover/cover_test.go:389]: no need to keep these

File: pkg/cover/instrument.go [hunk 6]
```diff
@@ -78,18 +73,12 @@ func loadValues() (map[string][]uint32, map[string][]testing.CoverBlock) {
...
 
 	{{range $file, $cover := .MainPkgCover.Vars}}
-	loadFileCover(coverCounters, coverBlocks, {{printf "%q" $cover.File}}, {{$cover.Var}}.Count[:], {{$cover.Var}}.Pos[:], {{$cover.Var}}.NumStmt[:])
-	{{end}}
-
-	{{range $k, $pkgCover := .CacheCover}}
-	{{range $v, $cover := $pkgCover.Vars}}
-	loadFileCover(coverCounters, coverBlocks, {{printf "%q" $cover.File}}, {{$pkgCover.Package.Name}}.{{$v}}.Count[:], {{$pkgCover.Package.Name}}.{{$v}}.Pos[:], {{$pkgCover.Package.Name}}.{{$v}}.NumStmt[:])
-	{{end}}
+	loadFileCover(coverCounters, coverBlocks, {{printf "%q" $cover.File}}, _cover.{{$cover.Var}}.Count[:], _cover.{{$cover.Var}}.Pos[:], _cover.{{$cover.Var}}.NumStmt[:])
```
Comment (CarlJi) [inline pkg/cover/instrument.go:81]: duplicate statement?

File: pkg/cover/instrument.go [hunk 7]
```diff
@@ -121,18 +110,12 @@ func clearValues() {
...
 
 	{{range $file, $cover := .MainPkgCover.Vars}}
-	clearFileCover({{$cover.Var}}.Count[:])
-	{{end}}
-
-	{{range $k, $pkgCover := .CacheCover}}
-	{{range $v, $cover := $pkgCover.Vars}}
-	clearFileCover({{$pkgCover.Package.Name}}.{{$v}}.Count[:])
-	{{end}}
+	clearFileCover(_cover.{{$cover.Var}}.Count[:])
```
Comment (CarlJi) [inline pkg/cover/instrument.go:118]: duplicate statement?

File: pkg/cover/instrument.go [hunk 8]
```diff
@@ -409,3 +392,18 @@ func checkCacheDir(p string) error {
...
+func injectGlobalCoverVarFile(ci *CoverInfo, content string) error {
+	coverFile, err := os.Create(filepath.Join(ci.Target, ci.GlobalCoverVarImportPath, "cover.go"))
+	if err != nil {
+		return err
+	}
+	defer coverFile.Close()
+
+	packageName := "package " + filepath.Base(ci.GlobalCoverVarImportPath) + "\n\n"
+
+	_, _ = coverFile.WriteString(packageName)
```
Comment (CarlJi) [inline pkg/cover/instrument.go:405]: it's not good practice to ignore error

File: pkg/cover/instrument.go [hunk 9]
```diff
@@ -409,3 +392,18 @@ func checkCacheDir(p string) error {
...
+	_, err = coverFile.WriteString(content)
```
Comment (CarlJi) [inline pkg/cover/instrument.go:406]: same with above

**Comment by CarlJi** [comment]
can we define the path `pkg/cover/cover2tool` like as `pkg/cover/internal/tool`？

File: pkg/cover/instrument.go (hunk 6 above)
Comment (lyyyuna) [reply pkg/cover/instrument.go:81]: main & deps' pkgs, not duplicate.

File: pkg/cover/instrument.go (hunk 7 above)
Comment (lyyyuna) [reply pkg/cover/instrument.go:118]: main & deps' pkgs, not duplicate.

**Pull request merged**